lib/helpers.py

```diff
@@ -10,437 +10,598 @@
 
-    deduplicated_mark_df = sorted_mark_df.drop_duplicates(['user_id'])
+        for chunk in pd.read_csv(filename, escapechar='\\', low_memory=False, **read_csv_kwargs):
+            if chunk_filter_fn:
+                filtered_chunk = chunk_filter_fn(chunk, self.revenue_event)
+            else:
+                filtered_chunk = chunk
 
-    return deduplicated_mark_df
+            if source != CSV_SOURCE_ATTRIBUTIONS:
+                # we are not interested in events that do not have a group amongst non-attribution events
+                filtered_chunk = filtered_chunk[filtered_chunk['ab_test_group'].isin(['test', 'control'])]
 
+            # remove events without a user id
+            filtered_chunk = filtered_chunk[filtered_chunk['user_id'].str.len() == 36]
 
-def merge(mark_df, attributions_df):
-    """
-    `merge` joins the marked users with the revenue events and excludes any revenue event that happened before the
-    user was marked.
-    """
-    merged_df = pd.merge(attributions_df, mark_df, on='user_id')
+            filtered_chunk = self._improve_types(filtered_chunk)
 
-    return merged_df[merged_df.ts_x > merged_df.ts_y]
+            df = pd.concat([df, filtered_chunk],
+                           ignore_index=True, verify_integrity=True)
 
+        print(datetime.now(), 'finished loading CSV for', date.strftime('%d.%m.%Y'),
+              'took', datetime.now() - now)
 
-def drop_duplicates_in_attributions(df, max_timedelta):
-    """
-    # Clean the data
+        if not os.path.exists(cache_dir):
+            os.makedirs(cache_dir)
 
-    Due to some inconsistencies in the measurement we need to clean the data before analysis.
+        print(datetime.now(), 'caching local as parquet', cache_file_name)
+        self._to_parquet(df, cache_file_name)
 
-    ### Remove duplicated events coming from AppsFlyer
+        # write it to the S3 cache folder as well
+        print(datetime.now(), 'caching on S3 as parquet', s3_cache_file_name)
+        self._to_parquet(df, s3_cache_file_name)
 
-    AppsFlyer is sending us two revenue events if they attribute the event to us. One of the events they send us does
-    not contain attribution information and the other one does. Sadly, it is not possible for us to distinguish
-    correctly if an event is a duplicate or if the user actually triggered two events with nearly the same information.
-    Therefore we rely on a heuristic. We consider an event a duplicate if the user and revenue are equal and the events
-    are less than a minute apart.
-    """
-    sorted = df.sort_values(['user_id', 'revenue_eur'])
+        return df
 
-    # Get values of the previous row
-    sorted['last_ts'] = sorted['ts'].shift(1)
-    sorted['last_user_id'] = sorted['user_id'].shift(1)
-    sorted['last_revenue'] = sorted['revenue_eur'].shift(1)
+    def _audience_data_path(self, audience):
+        return "s3://remerge-customers/{0}/uplift_data/{1}".format(self.customer, audience)
```

Comment:
Add S3 root path to constants?

Reply:
Why create an _extra_ named constant, when there's already a function, wrapping the said constant?

Reply:
Just to have a possibility to change S3 path from a single place, where all constants sit. But I agree, for this case it's unnecessary.